Sources/WalletConnectSign/Engine/Common/ApproveEngine.swift

```diff
@@ -233,74 +224,103 @@ private extension ApproveEngine {
+            pairingStore.delete(topic: payload.topic)
+        }
+        logger.debug("Session Proposal has been rejected")
+        kms.deletePrivateKey(for: payload.request.proposer.publicKey)
+
+        onSessionRejected?(
+            payload.request.publicRepresentation(),
+            SessionType.Reason(code: payload.error.code, message: payload.error.message)
+        )
+    }
+
     // MARK: SessionSettleResponse
 
-    func handleSessionSettleResponse(response: WCResponse) {
-        guard let session = sessionStore.getSession(forTopic: response.topic) else { return }
-        switch response.result {
-        case .response:
-            logger.debug("Received session settle response")
-            guard var session = sessionStore.getSession(forTopic: response.topic) else { return }
-            session.acknowledge()
-            sessionStore.setSession(session)
-        case .error(let error):
-            logger.error("Error - session rejected, Reason: \(error)")
-            networkingInteractor.unsubscribe(topic: response.topic)
-            sessionStore.delete(topic: response.topic)
-            kms.deleteAgreementSecret(for: response.topic)
-            kms.deletePrivateKey(for: session.publicKey!)
+    func handleSessionSettleResponse(payload: ResponseSubscriptionPayload<SessionType.SettleParams, Bool>) {
+        guard var session = sessionStore.getSession(forTopic: payload.topic) else {
+            return logger.debug(Errors.sessionNotFound.localizedDescription)
         }
+
+        logger.debug("Received session settle response")
+        session.acknowledge()
+        sessionStore.setSession(session)
+    }
+
+    func handleSessionSettleResponseError(payload: ResponseSubscriptionErrorPayload<SessionType.SettleParams>) {
+        guard let session = sessionStore.getSession(forTopic: payload.topic) else {
+            return logger.debug(Errors.sessionNotFound.localizedDescription)
+        }
+
+        logger.error("Error - session rejected, Reason: \(payload.error)")
+        networkingInteractor.unsubscribe(topic: payload.topic)
+        sessionStore.delete(topic: payload.topic)
+        kms.deleteAgreementSecret(for: payload.topic)
+        kms.deletePrivateKey(for: session.publicKey!)
     }
 
     // MARK: SessionProposeRequest
 
-    func handleSessionProposeRequest(payload: WCRequestSubscriptionPayload, proposal: SessionType.ProposeParams) throws {
+    func handleSessionProposeRequest(payload: RequestSubscriptionPayload<SessionType.ProposeParams>) {
         logger.debug("Received Session Proposal")
-        do { try Namespace.validate(proposal.requiredNamespaces) } catch { throw Errors.respondError(payload: payload, reason: .invalidUpdateRequest) }
+        let proposal = payload.request
+        do { try Namespace.validate(proposal.requiredNamespaces) } catch {
+            return respondError(payload: payload, reason: .invalidUpdateRequest, tag: SignProtocolMethod.sessionPropose.responseTag)
```

Comment:
I think the error reason is incorrect here

Reply:
ok, I can see you did not changed it. we can fix it later.
actually I see it is incorrect in other places as well